### PR TITLE
validation, refactor: UnloadBlockIndex and ChainstateManager::Reset thread safety cleanups

### PR DESCRIPTION
**Pull request by jonatack**
Thread safety refactoring seen in #24177:
- replace re-acquiring lock cs_main with a thread safety annotation in UnloadBlockIndex()
- remove ChainstateManager::Reset(), as it is currently unused (can be reintroduced in the test utilities if needed for unit testing)

### CODE REVIEW

File: src/validation.h [hunk 1]
```diff
@@ -985,17 +985,18 @@ class ChainstateManager
...
     //! Check to see if caches are out of balance and if so, call
     //! ResizeCoinsCaches() as needed.
     void MaybeRebalanceCaches() EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
 
     ~ChainstateManager() {
         LOCK(::cs_main);
-        UnloadBlockIndex(/* mempool */ nullptr, *this);
-        Reset();
+        UnloadBlockIndex(/*mempool=*/nullptr, *this);
+        // Clear (deconstruct) chainstate data.
```
[the remark this answers is not in the document]
Comment (jonatack) [reply src/validation.h:995]: Right! Removed (thanks!)

File: src/validation.cpp [hunk 2]
```diff
@@ -5009,15 +5009,6 @@ void ChainstateManager::Unload()
     m_best_invalid = nullptr;
 }
 
-void ChainstateManager::Reset()
```
Comment (jamesob) [inline src/validation.cpp:5012]: I make use of this in #24232

Comment (jonatack) [reply src/validation.cpp:5012]: Ok, reverted to the original version in 686d35de6 that keeps Reset() and replaces the lock with annotation+assertion.

Comment (MarcoFalke) [reply src/validation.cpp:5012]: (not saying what this pull should do, just saying what the long term goal should be)

**Review by vasild — APPROVED** [review]
ACK ae9ceed3e23288b163b7d7b1840b06b8d332f4ce

**Review by klementtan — APPROVED** [review]
crACK ae9ceed3e23288b163b7d7b1840b06b8d332f4ce

File: src/validation.cpp (hunk 2 above)
Comment (jamesob) [reply src/validation.cpp:5012]: This was not just "test-only code" - the Reset() was used to detect snapshot dirs that needed to be cleaned up. The merge here isn't now just a matter of copy/pasting some code.

**Pull request merged**